// @redhat-developer/vscode-redhat-telemetry/lib/common/vscode/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VSCodeSettings = void 0;
exports.getTelemetryConfiguration = getTelemetryConfiguration;
exports.isPreferenceOverridden = isPreferenceOverridden;
exports.didUserDisableTelemetry = didUserDisableTelemetry;
const vscode_1 = require("vscode");
const constants_1 = require("../impl/constants");
class VSCodeSettings {
    isTelemetryEnabled() {
        return this.getTelemetryLevel() !== 'off' && getTelemetryConfiguration().get('enabled', false);
    }
    getTelemetryLevel() {
        //Respecting old vscode telemetry settings https://github.com/microsoft/vscode/blob/f09c4124a229b4149984e1c2da46f35b873d23fa/src/vs/platform/telemetry/common/telemetryUtils.ts#L131
        if (vscode_1.workspace.getConfiguration().get("telemetry.enableTelemetry") == false
            || vscode_1.workspace.getConfiguration().get("telemetry.enableCrashReporter") == false) {
            return "off";
        }
        return vscode_1.workspace.getConfiguration().get("telemetry.telemetryLevel", "off");
    }
    isTelemetryConfigured() {
        return isPreferenceOverridden(constants_1.CONFIG_KEY + '.enabled');
    }
    updateTelemetryEnabledConfig(value) {
        return getTelemetryConfiguration().update('enabled', value, true);
    }
}
exports.VSCodeSettings = VSCodeSettings;
function getTelemetryConfiguration() {
    return vscode_1.workspace.getConfiguration(constants_1.CONFIG_KEY);
}
function isPreferenceOverridden(section) {
    const config = vscode_1.workspace.getConfiguration().inspect(section);
    return (config?.workspaceFolderValue !== undefined ||
        config?.workspaceFolderLanguageValue !== undefined ||
        config?.workspaceValue !== undefined ||
        config?.workspaceLanguageValue !== undefined ||
        config?.globalValue !== undefined ||
        config?.globalLanguageValue !== undefined);
}
function didUserDisableTelemetry() {
    if (vscode_1.env.isTelemetryEnabled) {
        return false;
    }
    //Telemetry is not enabled, but it might not be the user's choice.
    //i.e. could be the App's default setting (VS Codium), or  
    //then the user only asked for reporting errors/crashes, in which case we can do the same. 
    return isPreferenceOverridden("telemetry.telemetryLevel") && vscode_1.workspace.getConfiguration().get("telemetry.telemetryLevel") === "off";
}
